File: receiver.py
import time

class Receiver:
    def __init__(self, device, period, threshold,
                 start_pattern='11000', end_pattern='01101'):
        self.device = device
        self.period = period
        self.threshold = threshold
        self.start_pattern = start_pattern
        self.end_pattern = end_pattern

    def receive_bit(self):
        ns_in_s = 1000000000
        delta_ns = 1.05 * self.period - time.time_ns() % self.period
        time.sleep(delta_ns / ns_in_s)
        if self.device.read() >= self.threshold:
            return '1'
        return '0'

    def receive_bits(self, count):
        bits = ''
        for i in range(count):
            bits += self.receive_bit()
        return bits

    def receive(self):
        bits = ''
        while True:
            pattern = self.receive_bits(len(self.start_pattern))
            if pattern == self.end_pattern:
                return bits
            bits += pattern

File: receiver.c
#include "receiver.h"
#include <sys/time.h>
#include <unistd.h>
#include <stdlib.h>

struct lifi_receiver *lifi_receiver_new(void *device,
					void *read,
					unsigned long period,
					unsigned long threshold,
					size_t code_length,
					char start_code,
					char end_code)
{
	struct lifi_receiver *receiver = malloc(sizeof(*receiver));
	receiver->device = device;
	receiver->read = read;
	receiver->period = period;
	receiver->threshold = threshold;
	receiver->code_length = code_length;
	receiver->start_code = start_code;
	receiver->end_code = end_code;
	return receiver;
}

void lifi_receiver_delete(struct lifi_receiver *receiver)
{
	free(receiver);
}

char lifi_receive_bit(struct lifi_receiver *receiver)
{
	struct timeval current;
	gettimeofday(&current, NULL);
	usleep(receiver->period - current.tv_usec % receiver->period);
	if (receiver->read(receiver->device) > receiver->threshold)
		return 1;
	return 0;
}

char lifi_receive_bits(struct lifi_receiver *receiver, size_t count)
{
	char code = 0;
	for (size_t i = 0; i < count; ++i) {
		code <<= 1;
		code |= lifi_receive_bit(receiver);
	}
	return code;
}

int lifi_receive(struct lifi_receiver *receiver, char *str, size_t count)
{
	size_t i = 0;
	while (i < count) {
		char code = lifi_receive_bits(receiver, receiver->code_length);
		if (code == receiver->end_code)
			return i;
		str[i++] = code;
	}
	return -1;
}
